solidity/contracts/Gravity.sol

```diff
@@ -425,21 +442,45 @@ function updateValset(
 				// Send transaction amounts to destinations
 				uint256 totalFee;
 				for (uint256 i = 0; i < _amounts.length; i++) {
-					IERC20(_tokenContract).safeTransfer(_destinations[i], _amounts[i]);
+					transferNoRevert(_tokenContract, _destinations[i], _amounts[i]);
 					totalFee = totalFee + _fees[i];
 				}
 
 				// Send transaction fees to msg.sender
-				IERC20(_tokenContract).safeTransfer(msg.sender, totalFee);
+				transferNoRevert(_tokenContract, msg.sender, totalFee);
 			}
 		}
-
 		// LOGS scoped to reduce stack depth
 		{
 			state_lastEventNonce = state_lastEventNonce + 1;
 			emit TransactionBatchExecutedEvent(_batchNonce, _tokenContract, state_lastEventNonce);
 		}
-	}	
+	}
+
+	function transferNoRevert(address token, address to, uint value) internal {
+		try this.safeTransferSelf(token,to, value) {
+		} catch {
+			state_RevertedVouchers[state_lastRevertedNonce] = TransferReverted(token, to, value, true);
+			state_lastRevertedNonce = state_lastRevertedNonce + 1;
+		}
+	}
+
+	function safeTransferSelf(address token, address to, uint value) public onlySelf {
+		IERC20(token).safeTransfer(to, value);
+	}
+
+	function redeemVoucher(
+		uint256 _nonce,
+		address _newDestination
+	) public nonReentrant {
+		if(state_RevertedVouchers[_nonce].activate){
+			TransferReverted memory voucher = state_RevertedVouchers[_nonce];
+			require(voucher.destination == msg.sender);
+			IERC20(voucher.tokenContract).safeTransfer(_newDestination, voucher.amount);
```

Comment:
**check-effect design pattern**
Even though there is a reentrancyguard in place, I would suggest to still change the states first (amount = 0, activate=false) before doing the safeTransfer to prevent edge cases.

`state_RevertedVouchers[_nonce].activate = false;`
`state_RevertedVouchers[_nonce].amount = 0; `
`IERC20(voucher.tokenContract).safeTransfer(_newDestination, voucher.amount);`

Reply:
done